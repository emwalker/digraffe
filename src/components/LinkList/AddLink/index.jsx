// @flow
import React, { Component } from 'react'
import { graphql, createFragmentContainer } from 'react-relay'
import Select from 'react-select'
import { path } from 'ramda'

import createLinkMutation from '../../../mutations/createLinkMutation'
import selectTopicMutation from '../../../mutations/selectTopicMutation'
import { liftNodes } from '../../../utils'

const selectedTopic = path(['selectedTopic', 'value'])

type Props = {
  organization: {
    id: string,
    resourceId: string,
    availableTopics: Array<{
      label: string,
      value: string,
    }>,
  },
  viewer: {
    selectedTopic: {
      label: string,
      value: string,
    },
  },
  relay: {
    environment: Object,
  },
}

type State = {
  selectedValue: string,
  url: string,
}

class AddLink extends Component<Props, State> {
  constructor(props: Props) {
    super(props)
    this.options = liftNodes(props.organization.availableTopics)
    this.state = {
      selectedValue: selectedTopic(props.viewer),
      url: '',
    }
  }

  onInputChange = (event: Object) => {
    this.setState({ url: event.currentTarget.value })
  }

  onKeyPress = (event: Object) => {
    if (event.key === 'Enter')
      this.setState(({ url }) => {
        this.addLink(url)
        return { url: '' }
      })
  }

  onSelect = (option) => {
    const selectedValue = option ? option.value : null
    this.setState({ selectedValue }, () => {
      selectTopicMutation(
        this.props.relay.environment,
        {
          organizationId: this.props.organization.id,
          topicId: selectedValue || '',
        },
      )
    })
  }

  addLink = (url) => {
    const {
      id: orgId,
      resourceId: organizationId,
    } = this.props.organization
    const value = this.state.selectedValue
    const topicIds = value ? [value] : []

    createLinkMutation(
      this.props.relay.environment,
      orgId,
      {
        organizationId,
        url,
        topicIds,
      },
    )
  }

  options: Object[]

  render = () => (
    <div>
      <dl className="form-group">
        <Select
          id="link-topic-id"
          name="link-topic"
          onChange={this.onSelect}
          options={this.options}
          placeholder="Select a topic"
          value={this.state.selectedValue}
        />
      </dl>
      <dl className="form-group">
        <input
          className="form-control test-link-url"
          onChange={this.onInputChange}
          onKeyPress={this.onKeyPress}
          placeholder="Link URL"
          value={this.state.url}
        />
      </dl>
    </div>
  )
}

export default createFragmentContainer(AddLink, graphql`
  fragment AddLink_viewer on User {
    selectedTopic {
      label: name
      value: resourceId
    }
  }

  fragment AddLink_organization on Organization {
    id
    resourceId

    availableTopics: topics(first: 100) {
      edges {
        node {
          label: name
          value: resourceId
        }
      }
    }
  }
`)
