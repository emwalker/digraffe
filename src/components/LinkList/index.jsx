// @flow
import React from 'react'
import { graphql, createFragmentContainer } from 'react-relay'

import AddLink from './AddLink'
import ListView from '../ui/ListView'
import { liftNodes } from '../../utils'

type Props = {
  links: Object,
  organization: Object,
  relay: {
    environment: Object,
  },
  title: string,
  viewer: Object,
}

const LinkList = ({
  title, links, organization, relay, viewer,
}: Props) => (
  <ListView
    title={title}
    items={liftNodes(links)}
  >
    <AddLink
      className="test-add-link"
      organization={organization}
      relay={relay}
      viewer={viewer}
    />
  </ListView>
)

export default createFragmentContainer(LinkList, graphql`
  fragment LinkList_viewer on User {
    ...AddLink_viewer
  }

  fragment LinkList_organization on Organization {
    ...AddLink_organization
  }

  fragment LinkList_links on LinkConnection {
    edges {
      node {
        id
        display: title
        resourcePath

        topics(first: 5) {
          edges {
            node {
              name
              resourceId
              resourcePath
            }
          }
        }
      }
    }
  }
`)
