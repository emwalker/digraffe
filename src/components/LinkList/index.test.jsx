import React from 'react'
import { shallow } from 'enzyme'
import LinkList from './index'

jest.mock('react-relay', () => ({ createFragmentContainer: component => component }))

describe('<LinkList />', () => {
  const viewer = {
    name: 'Rezrov',
  }

  const organization = {}

  const relay = {
    environment: {},
  }

  const wrapper = shallow(
    <LinkList
      organization={organization}
      relay={relay}
      title="Frotz"
      viewer={viewer}
    />,
  )

  it('renders', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('includes a form to add a link', () => {
    expect(wrapper.find('.test-add-link')).toHaveLength(1)
  })
})
