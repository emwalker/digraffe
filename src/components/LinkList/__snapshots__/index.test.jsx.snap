// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LinkList /> renders 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LinkList
    organization={Object {}}
    relay={
      Object {
        "environment": Object {},
      }
    }
    title="Frotz"
    viewer={
      Object {
        "name": "Rezrov",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <AddLink
        className="test-add-link"
        organization={Object {}}
        relay={
          Object {
            "environment": Object {},
          }
        }
        viewer={
          Object {
            "name": "Rezrov",
          }
        }
      />,
      "items": Array [],
      "title": "Frotz",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "className": "test-add-link",
        "organization": Object {},
        "relay": Object {
          "environment": Object {},
        },
        "viewer": Object {
          "name": "Rezrov",
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <AddLink
          className="test-add-link"
          organization={Object {}}
          relay={
            Object {
              "environment": Object {},
            }
          }
          viewer={
            Object {
              "name": "Rezrov",
            }
          }
        />,
        "items": Array [],
        "title": "Frotz",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "className": "test-add-link",
          "organization": Object {},
          "relay": Object {
            "environment": Object {},
          },
          "viewer": Object {
            "name": "Rezrov",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
